File: frontend/lib/scoring.ts
// Ported from the former Python backend (services/scoring.py).
// Runs entirely in the browser — no server needed.

import type { Pole } from "@/lib/types";

const MBTI_TO_CHARACTER: Record<string, string> = {
  INFP: "Nyota", INFJ: "Nyota", ISFP: "Nyota", ISFJ: "Nyota",
  INTJ: "Hirono", INTP: "Hirono", ENTJ: "Hirono", ISTP: "Hirono",
  ENFP: "Lulu", ENTP: "Hirono", ENFJ: "Nyota", ESTP: "Lulu",
  ESFJ: "Lulu", ESFP: "Lulu", ESTJ: "Lulu", ISTJ: "Lulu",
};

// Ties default to I, N, F, P. (Each axis is asked an odd number of times, so
// with a full run a tie shouldn't actually occur — this is just a safety net.)
const TIE_DEFAULTS: Record<string, Pole> = {
  EI: "I", SN: "N", TF: "F", JP: "P",
};

/** Tally a list of chosen poles into a 4-letter type + its character. */
export function scoreFromPoles(poles: Pole[]): {
  type: string;
  character: string;
} {
  const count: Record<Pole, number> = {
    E: 0, I: 0, S: 0, N: 0, T: 0, F: 0, J: 0, P: 0,
  };
  for (const p of poles) count[p] += 1;

  const axes: [Pole, Pole, string][] = [
    ["E", "I", "EI"],
    ["S", "N", "SN"],
    ["T", "F", "TF"],
    ["J", "P", "JP"],
  ];

  let type = "";
  for (const [a, b, axis] of axes) {
    if (count[a] > count[b]) type += a;
    else if (count[b] > count[a]) type += b;
    else type += TIE_DEFAULTS[axis];
  }

  return { type, character: MBTI_TO_CHARACTER[type] ?? "Nyota" };
}

/** Slugify a figure name — must match the frontend convention. */
export function slugifyFigure(name: string): string {
  return name.toLowerCase().replace(/'/g, "").replace(/\s+/g, "-");
}

File: backend/app/services/scoring.py
MBTI_TO_CHARACTER = {
    "INFP": "Nyota", "INFJ": "Nyota", "ISFP": "Nyota", "ISFJ": "Nyota",
    "INTJ": "Hirono", "INTP": "Hirono", "ENTJ": "Hirono", "ISTP": "Hirono",
    "ENFP": "Lulu", "ENTP": "Hirono", "ENFJ": "Nyota", "ESTP": "Lulu",
    "ESFJ": "Lulu", "ESFP": "Lulu", "ESTJ": "Lulu", "ISTJ": "Lulu",
}

# Ties default to I, N, F, P
TIE_DEFAULTS = {"EI": "I", "SN": "N", "TF": "F", "JP": "P"}


def score_answers(answers: list[dict], questions_data: list[dict]) -> tuple[str, str]:
    """Return (mbti_type, character) from a list of {question_id, card_id} answers."""
    question_map = {q["id"]: q for q in questions_data}
    axis_scores: dict[str, dict[str, int]] = {
        "EI": {"E": 0, "I": 0},
        "SN": {"S": 0, "N": 0},
        "TF": {"T": 0, "F": 0},
        "JP": {"J": 0, "P": 0},
    }

    for answer in answers:
        q = question_map.get(answer["question_id"])
        if not q:
            continue
        card = next((c for c in q["cards"] if c["id"] == answer["card_id"]), None)
        if not card:
            continue
        axis = q["axis"]
        for pole, pts in card["scores"].items():
            axis_scores[axis][pole] += pts

    result = ""
    for axis, poles in axis_scores.items():
        p1, p2 = axis[0], axis[1]
        if poles[p1] > poles[p2]:
            result += p1
        elif poles[p2] > poles[p1]:
            result += p2
        else:
            result += TIE_DEFAULTS[axis]

    character = MBTI_TO_CHARACTER.get(result, "Nyota")
    return result, character
